File: script/generate.c
# Copyright (C) 2008-2010 Ali Gangji
# Distributed under the terms of the GNU General Public License v3
/**
 * This file is part of StarbugPHP
 * @file script/generate.c C wrapper for SAXON XSLT Processor that can be setuid to run generator from web interface
 * @author Ali Gangji <ali@neonrain.com>
 * @ingroup script
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
int main(int argc, char *argv[]) {
	if (strcmp(argv[1], "dir") == 0) {
		if (!file_exists(argv[2])) execl("/bin/mkdir", "/bin/mkdir", argv[2], (char *) 0);
	} else if (strcmp(argv[1], "sax") == 0) {
		printf("%s -jar %s %s %s %s", argv[2], argv[3], argv[4], argv[5], argv[6]);
		execl(argv[2], argv[2], "-jar", argv[3], argv[4], argv[5], argv[6], (char *) 0);
	} else if (strcmp(argv[1], "copy") == 0) {
		if (!file_exists(argv[2])) execl("/bin/cp", "/bin/cp", argv[3], argv[2], (char *) 0);
	}
}
int file_exists(const char * filename) {
	FILE * file = fopen(filename, "r");
	if (file != NULL) {
		fclose(file);
    return 1;
  }
  return 0;
}

File: script/generate.php
<?php
# Copyright (C) 2008-2010 Ali Gangji
# Distributed under the terms of the GNU General Public License v3
/**
 * This file is part of StarbugPHP <br/>
 * @file script/generate.php generates code from XSLT stylesheets
 * @author Ali Gangji <ali@neonrain.com>
 * @ingroup script
 */
global $sb;
// USAGE TEXT
$help = "Usage: generate TYPE NAME [OPTIONS]\n\n";
$help .= "TYPE\tWhat to generate\n";
$help .= "    \t\tcrud - CRUD views\n";
$help .= "    \t\tmodel - Object model\n";

$generator = array_shift($argv);
$model = array_shift($argv);

// IF GENERATING HOST, SKIP RIGHT TO IT AND EXIT
if ($generator == "host") include(dirname(__FILE__)."/generate/host/host.php");

// CLI VARS
assign("model", $model);
assign("generator", $generator);
$args = array();
foreach ($argv as $i => $arg) {
	if (0 === strpos($arg, "-")) {
		$arg = str_replace("-", "", $arg);
		$parts = (false !== strpos($arg, "=")) ? explode("=", $arg, 2) : array($arg, true);
		$args[$parts[0]] = $parts[1];
		assign($parts[0], $parts[1]);
	}
}

//SET VARS FOR GENERATOR 
$model_name = $model;
$dirs = array(); $generate = array(); $copy = array();

global $renderer;
$renderer->prefix = "script/generate/$generator/";

//SETUP XML
if ((!empty($model)) && (isset($schemer->tables[$model]))) {
	$data = $schemer->get($model);
	if (($generator == "model") || ($generator == "models")) {
		$schemer->toXML($data);
		$schemer->toJSON($data);
	}
}

//INCLUDE GENERATOR FILE
if (isset($args['u'])) include(BASE_DIR."/script/generate/$generator/update.php");
else include(BASE_DIR."/script/generate/$generator/$generator.php");

//CREATE DIRECTORIES
foreach ($dirs as $dir) if (!file_exists(BASE_DIR."/".$dir)) passthru("mkdir ".BASE_DIR."/$dir");
//CREATE FILES
foreach ($generate as $template => $output) {
	$o = BASE_DIR."/$output"; //output
	$data = capture($template);
	file_put_contents($o, $data);
}
//COPY FILES
foreach ($copy as $origin => $dest) passthru("cp ".BASE_DIR."/$dest ".BASE_DIR."/script/generate/$origin");
?>
